okio/src/test/java/okio/BufferedSourceTest.java

```diff
@@ -810,4 +810,109 @@ private void assertLongDecimalString(String s, long expected) throws IOException
+    List<ByteString> selection = Arrays.asList(a, bc, bd);
+
+    sink.write(bd);
+    sink.write(a);
+    sink.write(bc);
+
+    assertEquals(2, source.select(selection));
+    assertEquals(0, source.select(selection));
+    assertEquals(1, source.select(selection));
+    assertEquals(true, source.exhausted());
+  }
+
+  @Test public void selectNotFound() throws IOException {
+    List<ByteString> selection = Arrays.asList(
+        ByteString.encodeUtf8("ROCK"),
+        ByteString.encodeUtf8("SCISSORS"),
+        ByteString.encodeUtf8("PAPER"));
+
+    sink.writeUtf8("SPOCK");
+    assertEquals(-1, source.select(selection));
+    assertEquals("SPOCK", source.readUtf8());
+  }
+
+  @Test public void selectValuesHaveCommonPrefix() throws IOException {
+    List<ByteString> selection = Arrays.asList(
+        ByteString.encodeUtf8("abcd"),
+        ByteString.encodeUtf8("abce"),
+        ByteString.encodeUtf8("abcc"));
+
+    sink.writeUtf8("abcc").writeUtf8("abcd").writeUtf8("abce");
+    assertEquals(2, source.select(selection));
+    assertEquals(0, source.select(selection));
+    assertEquals(1, source.select(selection));
+  }
+
+  @Test public void selectLongerThanSource() throws IOException {
+    List<ByteString> selection = Arrays.asList(
+        ByteString.encodeUtf8("abcd"),
+        ByteString.encodeUtf8("abce"),
+        ByteString.encodeUtf8("abcc"));
+    sink.writeUtf8("abc");
+    assertEquals(-1, source.select(selection));
+    assertEquals("abc", source.readUtf8());
+  }
+
+  @Test public void selectReturnsFirstByteStringThatMatches() throws IOException {
+    List<ByteString> selection = Arrays.asList(
+        ByteString.encodeUtf8("abcd"),
+        ByteString.encodeUtf8("abc"),
+        ByteString.encodeUtf8("abcde"));
+    sink.writeUtf8("abcdef");
+    assertEquals(0, source.select(selection));
+    assertEquals("ef", source.readUtf8());
+  }
+
+  @Test public void selectNoByteStrings() throws IOException {
+    List<ByteString> selection = Arrays.asList();
+    sink.writeUtf8("abc");
+    assertEquals(-1, source.select(selection));
+  }
+
+  @Test public void selectFromEmptySource() throws IOException {
+    List<ByteString> selection = Arrays.asList(
+        ByteString.encodeUtf8("abc"),
+        ByteString.encodeUtf8("def"));
+    assertEquals(-1, source.select(selection));
+  }
+
+  @Test public void selectNoByteStringsFromEmptySource() throws IOException {
+    List<ByteString> selection = Arrays.asList();
+    assertEquals(-1, source.select(selection));
+  }
+
+  @Test public void selectEmptyByteString() throws IOException {
+    List<ByteString> selection = Arrays.asList(ByteString.of());
+    sink.writeUtf8("abc");
+    assertEquals(0, source.select(selection));
+    assertEquals("abc", source.readUtf8());
+  }
+
+  @Test public void selectEmptyByteStringFromEmptySource() throws IOException {
+    List<ByteString> selection = Arrays.asList(ByteString.of());
+    assertEquals(0, source.select(selection));
```

Comment:
What motivates this behavior? I'm not opposed to it, but it's different than I would have thought is desired.